### PR TITLE
Fixes to test suite to support CUDA arrays

### PR DESCRIPTION
**Pull request by gaurav-arya**
Implemented changes:

- A more generic construction of the random array in `test_plan_adjoint` 
- An option to disable testing wrapper types (in our case, views), since we do not support them correctly when it comes to avoiding ambiguities with `*` (see #119. I don't expect it to be solved in the immediate future, which is why I prefer to work around it for now).
- Tests correctness of `fftdims` as an iterable rather than its specific form, in accordance with the function contract.

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -5,6 +5,7 @@ version = "1.5.0"
 [deps]
 ChainRulesCore = "d360d2e6-b24c-11e9-a2a3-2a2ae2dbcce4"
 LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
+Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
```
Comment (gaurav-arya) [inline Project.toml:8]:
```suggestion
```

File: ext/AbstractFFTsTestExt.jl [hunk 2]
```diff
@@ -6,6 +6,7 @@ using AbstractFFTs
 using AbstractFFTs: TestUtils
 using AbstractFFTs.LinearAlgebra
 using Test
+import Random
```
Comment (gaurav-arya) [inline ext/AbstractFFTsTestExt.jl:9]:
```suggestion
```

File: ext/AbstractFFTsTestExt.jl [hunk 3]
```diff
@@ -70,9 +74,10 @@ function TestUtils.test_plan(P::AbstractFFTs.Plan, x::AbstractArray, x_transform
     end
 end
 
-function TestUtils.test_plan_adjoint(P::AbstractFFTs.Plan, x::AbstractArray; real_plan=false, copy_input=false)
+function TestUtils.test_plan_adjoint(P::AbstractFFTs.Plan, x::AbstractArray;
+                                     real_plan=false, copy_input=false, test_wrappers=true)
     _copy = copy_input ? copy : identity
-    y = rand(eltype(P * _copy(x)), size(P * _copy(x)))
+    y = Random.rand!(P * _copy(x))
```
Comment (gaurav-arya) [inline ext/AbstractFFTsTestExt.jl:80]:
```suggestion
    y = map(a -> rand(typeof(a)), P * _copy(x)) # generically construct rand array
```

**Review by gaurav-arya — COMMENTED** [review]
Remove `Random.rand!` usage

**Comment by gaurav-arya** [comment]
Bump on review:) It should hopefully be an easy review, especially because this does not touch the core functionality, only the test utilities, which I have verified continue to work with `FFTW` and after this also with `CUDA` on https://github.com/JuliaGPU/CUDA.jl/pull/2073

File: ext/AbstractFFTsTestExt.jl [hunk 4]
```diff
@@ -61,7 +62,9 @@ function TestUtils.test_plan(P::AbstractFFTs.Plan, x::AbstractArray, x_transform
         _x_out = similar(P * _copy(x))
         @test mul!(_x_out, P, _copy(x)) ≈ x_transformed
         @test _x_out ≈ x_transformed
-        @test P * view(_copy(x), axes(x)...) ≈ x_transformed # test view input
+        if test_wrappers
+            @test P * view(_copy(x), axes(x)...) ≈ x_transformed # test view input
+        end
```
Comment (devmotion) [inline ext/AbstractFFTsTestExt.jl:67]: It would be better to mark this as broken if it fails instead of disabling the test.

File: ext/AbstractFFTsTestExt.jl [hunk 5]
```diff
@@ -52,7 +52,8 @@ const TEST_CASES = (
         )
 
 
-function TestUtils.test_plan(P::AbstractFFTs.Plan, x::AbstractArray, x_transformed::AbstractArray; inplace_plan=false, copy_input=false)
+function TestUtils.test_plan(P::AbstractFFTs.Plan, x::AbstractArray, x_transformed::AbstractArray;
+                             inplace_plan=false, copy_input=false, test_wrappers=true)
```
Comment (devmotion) [inline ext/AbstractFFTsTestExt.jl:56]: Out of curiosity, why do we have `copy_input`? Couldn't we always copy the input? Or at least always copy when the plan is an in-place plan?

(It would also be good to deduce automatically whether a plan is in-place or not, in the worst case we could check whether `a = copy(x)` and `P * a` are aliased or modify the content of `a`.

File: ext/AbstractFFTsTestExt.jl [hunk 6]
```diff
@@ -71,9 +74,10 @@ function TestUtils.test_plan(P::AbstractFFTs.Plan, x::AbstractArray, x_transform
     end
 end
 
-function TestUtils.test_plan_adjoint(P::AbstractFFTs.Plan, x::AbstractArray; real_plan=false, copy_input=false)
+function TestUtils.test_plan_adjoint(P::AbstractFFTs.Plan, x::AbstractArray;
+                                     real_plan=false, copy_input=false, test_wrappers=true)
     _copy = copy_input ? copy : identity
-    y = rand(eltype(P * _copy(x)), size(P * _copy(x)))
+    y = map(a -> rand(typeof(a)), P * _copy(x)) # generically construct rand array
```
Comment (devmotion) [inline ext/AbstractFFTsTestExt.jl:80]: Can we use something more idiomatic such as
```suggestion
    P_x = P * _copy(x)
    y = rand!(similar(P_x))
```
(`similar` could be skipped if we are sure that `P_x` is mutable`)

File: ext/AbstractFFTsTestExt.jl (hunk 4 above)
Comment (gaurav-arya) [reply ext/AbstractFFTsTestExt.jl:67]: To clarify, if `test_wrappers = true`, we should use `@test`, and if `test_wrappers = false`, we should use `@test_broken`?

File: ext/AbstractFFTsTestExt.jl (hunk 5 above)
Comment (gaurav-arya) [reply ext/AbstractFFTsTestExt.jl:56]: This is because FFTW in some particular cases mutates the input even with an OOP plans (which has caused users some surprises). This is referenced in the docstring: https://github.com/JuliaMath/AbstractFFTs.jl/blob/a67bf157619b501b48954893b3a0044d6b8e6d0e/src/TestUtils.jl#L33-L34.

Since alternative OOP plan implementations might never mutate their input, they do not need to provide `copy_input=true` which makes for stronger testing of that property.

File: ext/AbstractFFTsTestExt.jl (hunk 6 above)
Comment (gaurav-arya) [reply ext/AbstractFFTsTestExt.jl:80]: I actually had this earlier, and removed it: https://github.com/JuliaMath/AbstractFFTs.jl/pull/118#pullrequestreview-1615665823

Although `map` and just using the eltype is clunkier, it doesn't assume mutability, and fits with the solution we came up with for mapping in the adjoint plans. Also, it avoids a dependency on `Random`, which would have to be depended on by `AbstractFFTs` itself so that it is accessible in the extension. Does that seem reasonable?

File: ext/AbstractFFTsTestExt.jl (hunk 4 above)
Comment (gaurav-arya) [reply ext/AbstractFFTsTestExt.jl]: One issue with `@test_broken` is that if the test fails sometimes, but not always, we get an unexpected pass error, which seems a bit brittle for users of `test_wrappers=false`. For now, I changed the test to use the skipping mechanism instead of my own branching, but let me know if you have further thoughts

File: ext/AbstractFFTsTestExt.jl (hunk 6 above)
Comment (gaurav-arya) [reply ext/AbstractFFTsTestExt.jl:78]: Oh, I see how `similar` allows us to supprt mutability. I think it's still a little finnicky because `y` is now a slightly different type, and there's also the `Random` dependency. Both should be fine for the FFTW and CUDA use cases, so I'm happy to go with what you think is best.

File: ext/AbstractFFTsTestExt.jl (hunk 5 above)
Comment (devmotion) [reply ext/AbstractFFTsTestExt.jl:56]: https://github.com/JuliaMath/AbstractFFTs.jl/issues/101 is a bug, isn't it? So I think the `copy_input = true` option should be removed.

Comment (gaurav-arya) [reply ext/AbstractFFTsTestExt.jl:56]: It's a behaviour of FFTW that's likely never going to change, see for example https://github.com/JuliaMath/FFTW.jl/issues/158 and https://github.com/JuliaMath/AbstractFFTs.jl/pull/88. Even CUFFT has it too I think, since past me added the option in https://github.com/JuliaGPU/CUDA.jl/pull/2073/files#diff-210fa8aa99b3f38d6045fb971b7ee0733923f8f14db526a29040d47d13b36f41R350 too. So I think we need `copy_input = true` to run meaningful tests in these cases

Comment (devmotion) [reply ext/AbstractFFTsTestExt.jl:56]: I don't see an argument for why it shouldn't be considered a bug in the linked issues and PRs?

File: ext/AbstractFFTsTestExt.jl (hunk 6 above)
Comment (devmotion) [reply ext/AbstractFFTsTestExt.jl:80]: The Random dependency would indeed be annoying. However, `map` could also lead to different types, it doesn't guarantee that `y` is of the same type as `P * x`.

To address your concerns, maybe one could use
```suggestion
    P_x = P * _copy(x)
    z = similar(P_x)
    Tz = eltype(z)
    @inbounds for i in eachindex(z)
        z[i] = rand(Tz)
    end
    y = convert(typeof(P_x), z)
```
`rand!` for generic arrays falls back to this loop anyway (with a `Sampler` which is a bit more efficient in general but does not really matter here IMO): https://github.com/JuliaLang/julia/blob/fb01dd28a5fbabdae6cb5f23c8493db9a377fda3/stdlib/Random/src/Random.jl#L269-L274
And the `convert` should address concerns about slightly different types and avoid e.g. changing from `StaticArrays.SArray` to `StaticArrays.MArray`.

File: ext/AbstractFFTsTestExt.jl (hunk 5 above)
Comment (gaurav-arya) [reply ext/AbstractFFTsTestExt.jl:56]: What do you propose as the path forward? It's a long-standing behaviour of the FFTW C library itself for a couple decades. Without fixing that behaviour we need the `copy_input` option . While the behaviour is unexpected (which is why `copy_input=false` is the default), practically, we need to support this case in the test set, as the only other way forward would be to somehow fix the C library. The behaviour cannot just be patched over by e.g. copying the input array in the plan application, because for many performance use cases the user may not care about the input array anymore after plan application.

It ought to be clearly documented both in `AbstractFFTs` (#88) and by downstream plan backends as a potential behaviour of some plans, which is a separate issue, but I don't see what other path forwards we have for this test set?

File: ext/AbstractFFTsTestExt.jl (hunk 6 above)
Comment (gaurav-arya) [reply ext/AbstractFFTsTestExt.jl:80]: I believe the scalar indexing might be problematic for CUDA arrays, at the very least causing warnings.

File: ext/AbstractFFTsTestExt.jl (hunk 5 above)
Comment (devmotion) [reply ext/AbstractFFTsTestExt.jl:56]: No, there's no need to fix or change anything in the FFTW library. FFTW.jl just has to copy the input array in this case: https://github.com/JuliaMath/FFTW.jl/issues/158#issuecomment-672244877

Comment (gaurav-arya) [reply ext/AbstractFFTsTestExt.jl:56]: I won't have the bandwidth to push forward that change forward in `FFTW` unfortunately. (My goal is ultimately just to fix the still-incorrect differentiation of `rfft` plans in `Zygote`.) Also, I think there might be pushback by others who believe the current behaviour should just be documented. Because if one allocates `y` and `x` and has the workflow
```julia
x = something
mul!(y, p, x);  
x = something else
mul!(y, p, x)
x = something else
```
then this code would already be safe and would now go from non-allocating to allocating. And even if it's a very non-ideal situation, presumably there are many reliances of this behaviour in the wild in a library as established as this one.

Comment (gaurav-arya) [reply ext/AbstractFFTsTestExt.jl:56]: I don't think that the mere presence of the `copy_input` argument, which is `false` by default, should make this PR blocked on handling an independent issue with FFTW.jl / CUFFT.jl that might take on the order of years to be addressed.

Comment (gaurav-arya) [reply ext/AbstractFFTsTestExt.jl:56]: I just noticed that there's already opposition to creating a copy: https://github.com/JuliaMath/FFTW.jl/issues/158#issuecomment-944397162. So we need to find a solution to this PR that doesn't require us tackling this issue -- is the `copy_input` option (which is `false` by default, only those backends with this mutating behaviour must explicitly set it `true`) acceptable to you? Without it, there is no way to run the adjoint plan tests on downstream trait implementations (which was the original purpose of these test utilities).

**Comment by gaurav-arya** [comment]
Bump on review. @devmotion raised in https://github.com/JuliaMath/AbstractFFTs.jl/pull/118#discussion_r1367955921) the issue of `map` not being contractually obligated to preserve the array type. But it does so in practice for all the cases I want to add these downstream tests, we didn't come up with a better solution, and we could easily improve the test suite with a backwards compatible solution in the future if that edge case does crop up.
